ext/LinearSolveEnzymeExt.jl

```diff
@@ -9,6 +9,47 @@ using Enzyme
 
 using EnzymeCore
 
+function EnzymeCore.EnzymeRules.forward(func::Const{typeof(LinearSolve.init)}, ::Type{RT}, prob::EnzymeCore.Annotation{LP}, alg::Const; kwargs...) where {RT, LP <: LinearSolve.LinearProblem}
+    @assert !(prob isa Const)
+    res = func.val(prob.val, alg.val; kwargs...)
+    if RT <: Const
+        return res
+    end
+    dres = func.val(prob.dval, alg.val; kwargs...)
+    dres.b .= res.b == dres.b ? zero(dres.b) : dres.b
+    dres.A .= res.A == dres.A ? zero(dres.A) : dres.A
+    if RT <: DuplicatedNoNeed
+        return dres
+    elseif RT <: Duplicated
+        return Duplicated(res, dres)
+    end
+end
+
+function EnzymeCore.EnzymeRules.forward(func::Const{typeof(LinearSolve.solve!)}, ::Type{RT}, linsolve::EnzymeCore.Annotation{LP}; kwargs...) where {RT, LP <: LinearSolve.LinearCache}
+    @assert !(linsolve isa Const)
+
+    A = deepcopy(linsolve.val.A) #mutates after function is applied
+    res = func.val(linsolve.val; kwargs...)
+    
+    if RT <: Const
+        return res
+    end
+    
+    dres = deepcopy(res)
+    invA = inv(A)
+    db = linsolve.dval.b
+    dA = linsolve.dval.A
+    dres.u .= invA * (db - dA * res.u)
```

Comment:
There's no need to refactorize here. This is just `A * u = db - dA * u`, or `u = A \ (db - dA * u)`. But this is just a `linsolve` call. Not only that, but it's the same operator as the one used in the forward pass, so you don't need to refactorize `A`. Therefore, this should simply use the same `linsolve` and do `linsolve.b = db - dA * u` and then `solve!`.

But I don't think the formula is correct. Isn't it just `linsolve.b = db` and then `solve!(linsolve)` then `du = linsolve.u`?

Reply:
But `linsolves.A` seem to mutate after the solve. That is why I had to add all this other stuff.

MWE:
```julia
using LinearSolve
A = rand(5,5)
b = rand(5)

prob = LinearProblem(A,b)
linsolve = init(prob)
@show linsolve.A
sol = solve!(linsolve)
@show linsolve.A
```
```
5×5 Matrix{Float64}:
 0.207897  0.0737386  0.220551   0.935437  0.883482
 0.362979  0.687049   0.220086   0.216771  0.145246
 0.538937  0.577133   0.965879   0.438704  0.689019
 0.348145  0.174776   0.0491639  0.145817  0.155529
 0.92258   0.637729   0.508441   0.166917  0.218296

5×5 Matrix{Float64}:
 0.92258    0.637729   0.508441   0.166917  0.218296
 0.393439   0.436142   0.0200455  0.1511    0.0593595
 0.584162   0.469106   0.659464   0.270316  0.533653
 0.225343  -0.160427   0.16558    0.877305  0.755451
 0.37736   -0.151045  -0.211799   0.185687  0.0548682
```

Reply:
Yes, the mutation that it's doing is `lu!(A)`, i.e. its using the same memory in the representation of the LU-factorization. You want to let it use exactly the same mutated `A` via `cache.cacheval` in order to do the next solve, which is precisely what's done in the caching interface. See https://docs.sciml.ai/LinearSolve/stable/tutorials/caching_interface/

Reply:
Sounds good. And the simple  derivation of the formula is mentioned in https://scicomp.stackexchange.com/a/29421.

Reply:
Also, the correctness of the formula is checked with `FiniteDiff` in the tests - https://github.com/SciML/LinearSolve.jl/pull/416/files#diff-7c5a302dc0e55153407f2354959d201e28d2b9a61c4ef6bc86f05697a52b4cacR180-R196

Reply:
addressed in 31745cc